docs/references/ingress-messages.mdx

```diff
@@ -0,0 +1,94 @@
+
+ICP provides a built-in deduplication feature. Namely, if two calls with identical [request id](/docs/current/references/ic-interface-spec#request-ids) are made, the message is processed at most once. For two calls to have the same `request_id`, they must have the same values for all fields `request_type`, `sender`, `nonce`, `canister_id`, `method_name`, `arg` and **ingress_expiry**. This gives idempotency, but only until the [`ingress_expiry`](/docs/current/references/ic-interface-spec#authentication) is reached, at which point the message expires and ICP [may remove the `request_id` from the system state tree](/docs/current/references/ic-interface-spec#state-tree-request-status). This isn't an issue since the same request will not be accepted later, as it expired. Thus, such requests can be safely retried analogously to how it is done for [idempotent calls](/docs/current/developer-docs/smart-contracts/best-practices/idempotency#idempotent-canister-apis), see the illustration on that page.
+
+As ICP rejects `ingress_expiry` that is too far in the future, the time window for retrying is limited. To maximize the window, one can maximize `ingress_expiry` in this case and time out the retries after `ingress_expiry` as well.
+
+In summary, there are two options for a client:
+* _No retries, only poll `read_state`:_ Submit the `call` once, and poll `read_state` until at least one of the following conditions is met:
+  * A certified response was produced. That is, the call status is `replied`, `rejected`, or `done`.
+  * `ingress_expiry` is reached, and the call status is not `received` or `processing`.
+  
+  Evaluating the second condition may be hard if the subnet time and local time of the user differ. 
+* _Retry the call while polling `read_state`_ Do the same as above, but resubmit the initial `call` (same `request_id`) periodically (e.g. every `30s`) until at least one of the following conditions is met: 
+  * The call is processing or a certified response was produced. That is, the call status is `received`, `processing`, `replied`, `rejected`, or `done`.
+  * `ingress_expiry` is reached.
+
+  This can be beneficial because a call that didn’t make it is retried. For example, if the first call was dropped due to a network issue or by a malicious replica, a subsequent call could make it through e.g. if it is successfully routed to an honest replica. 
+
+At the time of writing, both agent-js (v2.1.3) and agent-rs (v0.39.1) don't retry `call`, so they follow the first option above. By default, agent-js uses `ingress_expiry` of five minutes and also polls for five minutes. agent-rs sets a default `ingress_expiry` of four minutes and polls for five minutes, but whenever it observes a `received` or `processing` status, it resets the timer to allow for an additional four minutes of polling. 
```

Comment:
agent-js uses four minutes as well (5min - 1min) https://github.com/dfinity/agent-js/blob/main/packages/agent/src/agent/http/transforms.ts#L31 when setting the expiry. I don't know where in the code the polling is stopped.

Reply:
Oh good catch, thanks. I missed the additional logic in the constructor. Adapted to 4 min. 

> I don't know where in the code the polling is stopped.

FYI this is done [here](https://github.com/dfinity/agent-js/blob/12b851f01fb0a0fc911089f37282adcb59e991bb/packages/agent/src/polling/index.ts#L28). [This is the default strategy.](https://github.com/dfinity/agent-js/blob/12b851f01fb0a0fc911089f37282adcb59e991bb/packages/agent/src/polling/strategy.ts#L19-L21)